### PR TITLE
[UX] Improve ux when nixpkgs has not been downloaded

### PR DESCRIPTION
**Pull request by mikeland86**
## Summary

Ensure that nixkgs has been downloaded in a few places we were expecting it and causing poor UX.

## How was it tested?

* Installed a few packages
* unit tests

### CODE REVIEW

File: internal/impl/packages.go [hunk 1]
```diff
@@ -219,6 +219,15 @@ func (d *Devbox) addPackagesToProfile(ctx context.Context, mode installMode) err
...
 	}
 
+	// If packages are in profile but nixpkgs has been purged, the experience
+	// will be poor when we try to run print-dev-env. So we ensure nixpkgs is
+	// prefetched for all our packages.
+	for _, input := range d.packagesAsInputs() {
+		if err := input.EnsureNixpkgsPrefetched(d.writer); err != nil {
+			return err
+		}
+	}
```
[the remark this answers is not in the document]
Comment (savil) [reply internal/impl/packages.go:229]: Greg clarified offline that nixpkgs is not normally considered a dependency

File: internal/nix/input_test.go [hunk 2]
```diff
@@ -142,3 +142,47 @@ func (l *lockfile) Resolve(pkg string) (*lock.Package, error) {
...
+		},
+		{
+			url:      "github:NixOS/other-repo/12345",
+			expected: "",
+		},
+		{
+			url:      "",
+			expected: "",
+		},
+	}
```
Comment (savil) [inline internal/nix/input_test.go:175]: thanks for adding these test cases

File: internal/nix/search.go [hunk 3]
```diff
@@ -94,6 +94,14 @@ func pkgExistsForAnySystem(pkg string) bool {
 }
 
 func searchSystem(url string, system string) map[string]*Info {
+	// Search will download nixpkgs if it's not already downloaded. Adding this
+	// check here provides a slightly better UX.
+	if IsGithubNixpkgsURL(url) {
+		hash := HashFromNixPkgsURL(url)
+		// TODO, pass legit writer, handle error
```
Comment (savil) [inline internal/nix/search.go:101]: could we fix this prior to landing, otherwise we'll likely forget

**Review by savil — APPROVED** [review]
cool

Does this impact `devbox shell` start time?

File: internal/nix/search.go (hunk 3 above)
Comment (mikeland86) [reply internal/nix/search.go]: I added some comments on why stderr is OK and why we want to ignore the error. Passing in a writer is a bit complicated. I think longer term we can restructure the nix package to have an struct. Another option is to hard code everything to stderr.

**Pull request merged**